### PR TITLE
 Update STL reader.

### PR DESCRIPTION
**Pull request by jdumas**
The following STLs from Thingi10K were not loading with the previous STL parser:
- 1482988.stl
- 1080516.stl
- 56117.stl

This PR updates our STL parser (copied over from [PyMesh](https://github.com/PyMesh/PyMesh)) to fix those issues. Previous code had a lot of `goto` in it so I got scared, and preferred to use another implementation.

#### Check all that apply (change to `[x]`)
- [x] All changes meet [libigl style-guidelines](https://libigl.github.io/style-guidelines/).
- [x] Adds new .cpp file.
- [ ] Adds corresponding unit test.
- [ ] This is a minor change.

### CODE REVIEW

File: include/igl/list_to_matrix.cpp [hunk 1]
```diff
@@ -52,6 +52,39 @@ IGL_INLINE bool igl::list_to_matrix(const std::vector<std::vector<T > > & V,Eige
...
+{
+  // number of rows
+  int m = V.size();
+  if(m == 0)
+  {
+    M.resize(
+            Derived::RowsAtCompileTime>=0?Derived::RowsAtCompileTime:0
+            ,
+            Derived::ColsAtCompileTime>=0?Derived::ColsAtCompileTime:0
+            );
```
Comment (qnzhou) [inline include/igl/list_to_matrix.cpp:66]: Clang-format can be pretty bad sometimes.

File: include/igl/readSTL.cpp [hunk 2]
```diff
@@ -1,290 +1,322 @@
...
-      is_ascii = true;
-    }
+}
+
+template <typename TypeV, typename TypeF, typename TypeN>
+IGL_INLINE bool read_stl_ascii(std::istream &input,
+                               std::vector<std::array<TypeV, 3>> &V,
+                               std::vector<std::array<TypeF, 3>> &F,
+                               std::vector<std::array<TypeN, 3>> &N) {
+  const size_t LINE_SIZE = 256;
```
Comment (qnzhou) [inline include/igl/readSTL.cpp:79]: Same here.

File: include/igl/readSTL.cpp [hunk 3]
```diff
@@ -1,290 +1,322 @@
...
-  std::vector<std::vector<TypeV> > & V,
-  std::vector<std::vector<TypeF> > & F,
-  std::vector<std::vector<TypeN> > & N)
-{
-  using namespace std;
-  //stl_file = freopen(NULL,"rb",stl_file);
-  if(NULL==stl_file)
-  {
-    fprintf(stderr,"IOError: stl file could not be reopened as binary (1) ...\n");
+  const size_t HEADER_SIZE = 80;
```
Comment (qnzhou) [inline include/igl/readSTL.cpp:48]: Same here.

File: include/igl/readSTL.cpp [hunk 4]
```diff
@@ -1,290 +1,322 @@
...
-  using namespace std;
-  vector<vector<typename DerivedV::Scalar> > vV;
-  vector<vector<typename DerivedN::Scalar> > vN;
-  vector<vector<typename DerivedF::Scalar> > vF;
-  if(!readSTL(filename,vV,vF,vN))
-  {
+IGL_INLINE bool readSTL(std::istream &input,
+                        Eigen::PlainObjectBase<DerivedV> &V,
+                        Eigen::PlainObjectBase<DerivedF> &F,
+                        Eigen::PlainObjectBase<DerivedN> &N) {
```
Comment (qnzhou) [inline include/igl/readSTL.cpp:23]: I thought libigl is moving to `Eigen::MatrixBase`?

**Review by qnzhou — APPROVED** [review]
Thanks Jérémie!  Looks good to me!  Just minor comments.

File: include/igl/readSTL.cpp (hunk 4 above)
Comment (jdumas) [reply include/igl/readSTL.cpp:23]: Those are output args, so they should be plain objects.

File: include/igl/list_to_matrix.cpp (hunk 1 above)
Comment (jdumas) [reply include/igl/list_to_matrix.cpp:66]: I gave up on code style for libigl as a whole. Maybe one day we'll adopt a .clang-format file...

**Pull request merged**